src/app/clusters/time-synchronization-server/time-synchronization-server.cpp

```diff
@@ -0,0 +1,980 @@
+        // list should be sorted by validStarting
+        // validUntil shall be larger than validStarting
+        if (!dstItem.validUntil.IsNull() && dstItem.validStarting >= dstItem.validUntil.Value())
+        {
+            mTimeSyncDataProvider.LoadDSTOffset(mDstOffsetObj);
+            return CHIP_ERROR_INVALID_TIME;
+        }
+        // validStarting shall not be smaller than validUntil of previous entry
+        if (dstItem.validStarting < lastValidUntil)
+        {
+            mTimeSyncDataProvider.LoadDSTOffset(mDstOffsetObj);
+            return CHIP_ERROR_INVALID_TIME;
+        }
+        lastValidUntil = !dstItem.validUntil.IsNull() ? dstItem.validUntil.Value() : lastValidUntil;
+        // only 1 validUntil null value and shall be last in the list
+        if (dstItem.validUntil.IsNull() && (i != mDstOffsetObj.size - 1))
+        {
+            mTimeSyncDataProvider.LoadDSTOffset(mDstOffsetObj);
+            return CHIP_ERROR_INVALID_TIME;
+        }
+    }
+
+    return mTimeSyncDataProvider.StoreDSTOffset(GetDSTOffset());
+}
+
+CHIP_ERROR TimeSynchronizationServer::ClearTimeZone()
+{
+    mTimeZoneObj.size         = 1; // one default time zone item is needed
+    mTimeZoneObj.timeZoneList = Span<TimeZoneStore>(mTz);
+    for (auto & tzObj : mTimeZoneObj.timeZoneList)
+    {
+        memset(tzObj.name, 0, sizeof(tzObj.name));
+        tzObj.timeZone = { .offset = 0, .validAt = 0, .name = MakeOptional(CharSpan(tzObj.name, sizeof(tzObj.name))) };
+    }
+    return mTimeSyncDataProvider.StoreTimeZone(GetTimeZone());
+}
+
+CHIP_ERROR TimeSynchronizationServer::ClearDSTOffset()
+{
+    mDstOffsetObj.size = 0;
+    return mTimeSyncDataProvider.ClearDSTOffset();
+}
+
+DataModel::Nullable<Structs::TrustedTimeSourceStruct::Type> & TimeSynchronizationServer::GetTrustedTimeSource()
+{
+    return mTrustedTimeSource;
+}
+
+CHIP_ERROR TimeSynchronizationServer::GetDefaultNtp(MutableCharSpan & dntp)
+{
+    return mTimeSyncDataProvider.LoadDefaultNtp(dntp);
+}
+
+Span<TimeZoneStore> TimeSynchronizationServer::GetTimeZone()
+{
+    return mTimeZoneObj.timeZoneList.SubSpan(0, mTimeZoneObj.size);
+}
+
+DataModel::List<Structs::DSTOffsetStruct::Type> TimeSynchronizationServer::GetDSTOffset()
+{
+    return mDstOffsetObj.dstOffsetList.SubSpan(0, mDstOffsetObj.size);
+}
+
+void TimeSynchronizationServer::ScheduleDelayedAction(System::Clock::Seconds32 delay, System::TimerCompleteCallback action,
+                                                      void * aAppState)
+{
+    if (CHIP_NO_ERROR != SystemLayer().StartTimer(std::chrono::duration_cast<System::Clock::Timeout>(delay), action, aAppState))
+    {
+        ChipLogError(Zcl, "Time Synchronization failed to schedule timer.");
+    }
+}
+
+CHIP_ERROR TimeSynchronizationServer::SetUTCTime(EndpointId ep, uint64_t utcTime, GranularityEnum granularity,
+                                                 TimeSourceEnum source)
+{
+    ReturnErrorOnFailure(UpdateUTCTime(utcTime));
```

Comment:
Somewhere in here we should be adjusting what set of timezones/DST-offsets/etc we are using, no?

Which brings up another issue: we are currently dropping things from those lists in various cases, but we could be dropping them based on incorrect times....

Followup for this, obviously; this will take some thoughts to get right.

Reply:
Tracked in https://github.com/project-chip/connectedhomeip/issues/27575, resolving this comment for now, further followup on the issue.